common/reactive/src/main/java/io/helidon/common/reactive/BufferedEmittingPublisher.java

```diff
@@ -0,0 +1,299 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.common.reactive;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Flow;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Emitting publisher for manual publishing with built-in buffer for handling backpressure.
+ *
+ * <p>
+ * <strong>This publisher allows only a single subscriber</strong>.
+ * </p>
+ *
+ * @param <T> type of emitted item
+ */
+public class BufferedEmittingPublisher<T> implements Flow.Publisher<T> {
+
+    private final AtomicReference<State> state = new AtomicReference<>(State.READY_TO_EMIT);
+    private final ConcurrentLinkedQueue<T> buffer = new ConcurrentLinkedQueue<>();
+    private final EmittingPublisher<T> emitter = new EmittingPublisher<>();
+    private final AtomicBoolean draining = new AtomicBoolean(false);
+    private final AtomicBoolean emitting = new AtomicBoolean(false);
+    private final AtomicReference<Throwable> error = new AtomicReference<>();
+    private BiConsumer<Long, Long> requestCallback = (n, r) -> {};
+
+    protected BufferedEmittingPublisher() {
+    }
+
+    /**
+     * Create new {@link BufferedEmittingPublisher}.
+     *
+     * @param <T> type of emitted item
+     * @return new instance of BufferedEmittingPublisher
+     */
+    public static <T> BufferedEmittingPublisher<T> create() {
+        return new BufferedEmittingPublisher<T>();
+    }
+
+    @Override
+    public void subscribe(final Flow.Subscriber<? super T> subscriber) {
+        emitter.onSubscribe(() -> state.get().drain(this));
+        emitter.onRequest((n, cnt) -> {
+            requestCallback.accept(n, cnt);
+            state.get().drain(this);
+        });
+        emitter.onCancel(() -> state.compareAndSet(State.READY_TO_EMIT, State.CANCELLED));
+        emitter.subscribe(subscriber);
+    }
+
+    /**
+     * Callback executed when request signal from downstream arrive.
+     * <ul>
+     * <li><b>param</b> {@code n} the requested count.</li>
+     * <li><b>param</b> {@code result} the current total cumulative requested count, ranges between [0, {@link Long#MAX_VALUE}]
+     * where the max indicates that this publisher is unbounded.</li>
+     * </ul>
+     *
+     * @param requestCallback to be executed
+     */
+    public void onRequest(BiConsumer<Long, Long> requestCallback) {
+        this.requestCallback = BiConsumerChain.combine(this.requestCallback, requestCallback);
+    }
+
```

Comment:
How about having support for onEmit callbacks ?

Reply:
Would be sufficient to invoke such callback right after emit invocation? To keep request counting isolated in `EmittingPublisher` the `BufferedEmittingPublisher` never know in advance if emit will be successful

Reply:
What you did seems sufficient to me.